pkg/operator/starter.go

```diff
@@ -50,62 +52,95 @@ const (
+	// Create core clientset for the GUEST cluster.
+	guestNamespace := defaultNamespace
+	guestKubeConfig := controllerConfig.KubeConfig
+	guestKubeClient := controlPlaneKubeClient
+	isHypershift := *guestKubeConfigString != ""
+	if isHypershift {
+		guestKubeConfig, err = client.GetKubeConfigOrInClusterConfig(*guestKubeConfigString, nil)
+		if err != nil {
+			return err
+		}
+		guestKubeClient = kubeclient.NewForConfigOrDie(rest.AddUserAgent(guestKubeConfig, operatorName))
+
+		// Create all events in the GUEST cluster.
+		// Use name of the operator Deployment in the management cluster + namespace
+		// in the guest cluster as the closest approximation of the real involvedObject.
+		controllerRef, err := events.GetControllerReferenceForCurrentPod(ctx, controlPlaneKubeClient, controlPlaneNamespace, nil)
+		controllerRef.Namespace = guestNamespace
+		if err != nil {
+			klog.Warningf("unable to get owner reference (falling back to namespace): %v", err)
+		}
+		eventRecorder = events.NewKubeRecorder(guestKubeClient.CoreV1().Events(guestNamespace), operandName, controllerRef)
+	}
+
+	guestDynamicClient, err := dynamic.NewForConfig(guestKubeConfig)
 	if err != nil {
 		return err
 	}
 
-	csiControllerSet := csicontrollerset.NewCSIControllerSet(
-		operatorClient,
-		controllerConfig.EventRecorder,
+	// Client informers for the GUEST cluster.
+	guestKubeInformersForNamespaces := v1helpers.NewKubeInformersForNamespaces(guestKubeClient, guestNamespace, "")
+	guestConfigMapInformer := guestKubeInformersForNamespaces.InformersFor(guestNamespace).Core().V1().ConfigMaps()
+	guestNodeInformer := guestKubeInformersForNamespaces.InformersFor("").Core().V1().Nodes()
+
+	guestConfigClient := configclient.NewForConfigOrDie(rest.AddUserAgent(guestKubeConfig, operatorName))
+	guestConfigInformers := configinformers.NewSharedInformerFactory(guestConfigClient, 20*time.Minute)
+	guestInfraInformer := guestConfigInformers.Config().V1().Infrastructures()
+
+	// Create client and informers for our ClusterCSIDriver CR.
+	gvr := opv1.SchemeGroupVersion.WithResource("clustercsidrivers")
+	guestOperatorClient, guestDynamicInformers, err := goc.NewClusterScopedOperatorClientWithConfigName(guestKubeConfig, gvr, string(opv1.AWSEBSCSIDriver))
+	if err != nil {
+		return err
+	}
+
+	// Start controllers that manage resources in the MANAGEMENT cluster.
+	controlPlaneCSIControllerSet := csicontrollerset.NewCSIControllerSet(
+		guestOperatorClient,
+		eventRecorder,
 	).WithLogLevelController().WithManagementStateController(
 		operandName,
 		false,
 	).WithStaticResourcesController(
 		"AWSEBSDriverStaticResourcesController",
-		kubeClient,
-		dynamicClient,
-		kubeInformersForNamespaces,
-		assets.ReadFile,
+		controlPlaneKubeClient,
+		controlPlaneDynamicClient,
+		controlPlaneKubeInformersForNamespaces,
+		assetWithNamespaceFunc(controlPlaneNamespace),
 		[]string{
 			"storageclass_gp2.yaml",
 			"volumesnapshotclass.yaml",
 			"csidriver.yaml",
```

Comment:
These three should be in the guest cluster.

Reply:
Fixed.